db/db_impl/db_impl_open.cc

```diff
@@ -614,14 +614,16 @@ Status DBImpl::Recover(
     // read-write db instance because options_file_number_ will later be
     // updated to versions_->NewFileNumber() in RenameTempFileToOptionsFile.
     std::vector<std::string>* filenames;
-    if (s.ok() && dbname_ != immutable_db_options_.wal_dir) {
-      // GetChildren() on dbname_ was NOT called above.
-      s = env_->GetChildren(dbname_, &files_in_dbname);
-      filenames = &files_in_dbname;
-    } else {
-      s = Status::OK();
-      filenames = &files_in_wal_dir;
-    }
+    if (s.ok()) {
+      if (dbname_ != immutable_db_options_.wal_dir){
+        // GetChildren() on dbname_ was NOT called above.
+        s = env_->GetChildren(dbname_, &files_in_dbname);
+        filenames = &files_in_dbname;
+      } else {
+        s = Status::OK();
+        filenames = &files_in_wal_dir;
+      }
+    } 
```

Comment:
I don't feel so comfortable here... any thoughts?

Reply:
What is the concern?

Reply:
It seems that it just moved the pointer juggling from `wal_dir`-related section to here :) anyways I'm OK with it as long as it looks good to you.

Reply:
When we call `GetChildren()` on `wal_dir` for the first time, we know it must be empty since `best_efforts_recovery` must be false, and we must not have called `GetChildren(dbname)` at all. Therefore, there is no need to check whether `dbname_ == wal_dir`. Even if they are equal, we need to call `GetChildren(wal_dir)`.

In the final `read_only` if block, we do not have to call `GetChildren()` on `dbname` again if
- best_efforts_recovery is true, or
- best_efforts_recovery is false and dbname_ is equal to wal_dir.

Reply:
Gotcha. Thanks a lot!

Reply:
> * best_efforts_recovery is true, or

This is assuming we called `getChildren()` at `TryRecover()`, right? If that's the case, we need some way to cache the `getChildren()` result in `ManifestPicker`. I think we need to roll back to the older version where `files_in_dbname` is a member in `ManifestPicker` so that we can have the result returned.

Reply:
We can let `TryRecover()` return the result of `GetChildren()`. For example (without considering error handling for simplicity). Still no need to add member variable `file_names_` to `ManifestPicker`.
```
TryRecovery(......, bool read_only, ......, std::vector<std::string>& files_in_dbname) {
  if (read_only) {
    assert(files_in_dbname.empty());
    fs_->GetChildren(&files_in_dbname);
  }
  ManifestPicker(..., files_in_dbname);
  ...
}
```

Reply:
Or just do a `GetChildren()` if `best_efforts_recovery && read_only`.